From e7c03c2f942e5bd16e859e24b43ce6aa357adc9f Mon Sep 17 00:00:00 2001
From: Thang Tran <thuutran@amperecomputing.com>
Date: Thu, 21 Nov 2024 07:09:02 +0000
Subject: [PATCH 6/6] psusensor: monitor interfaceAdded signal for CPU object

When the phosphor-inventory creates the object path, it does not send
"PropertiesChanged" signal, it sends "interfacesAdded" instead. It makes
the psusensor can not know object paths that are created psusensor has
been started.

Solution: Monitor "interfacesAdded" signals to catch object path that
presenst CPU status.

Tested:
     PSU sensors that depend on the CPUIndex property in Entity-Manager
     are reported correctly.

Upstream-Status: Pending

Signed-off-by: Thang Tran <thuutran@amperecomputing.com>
---
 src/psu/PSUSensorMain.cpp | 60 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 60 insertions(+)

diff --git a/src/psu/PSUSensorMain.cpp b/src/psu/PSUSensorMain.cpp
index 3f9997c..53d8164 100644
--- a/src/psu/PSUSensorMain.cpp
+++ b/src/psu/PSUSensorMain.cpp
@@ -1255,6 +1255,61 @@ int main()
             });
         };

+    std::function<void(sdbusplus::message_t&)> addCpuPresenceHandler =
+        [&](sdbusplus::message_t& message) {
+            sdbusplus::message::object_path cpuPath;
+            std::map<std::string, std::map<std::string, BasicVariantType>>
+                interfaces;
+            message.read(cpuPath, interfaces);
+            std::string cpuName = cpuPath.filename();
+            boost::to_lower(cpuName);
+
+            if (!cpuName.starts_with("cpu"))
+            {
+                return; // not interested
+            }
+
+            size_t index = 0;
+            try
+            {
+                index = std::stoi(cpuName.substr(cpuName.size() - 1));
+            }
+            catch (const std::invalid_argument&)
+            {
+                std::cerr << "Found invalid path " << cpuInventoryPath << "/"
+                            << cpuName << "\n";
+                return;
+            }
+
+            /* This message is sent by application that include Inventory.Item
+             * interface, therefore We don't have to check finding Inventory.Item
+             * interface is success or not */
+            const auto& values = interfaces[inventoryItemIntf];
+            const auto& findPresence = values.find("Present");
+
+            if (findPresence != values.end())
+            {
+                cpuPresence[index] = std::get<bool>(findPresence->second);
+            }
+
+            // this implicitly cancels the timer
+            cpuFilterTimer.expires_after(std::chrono::seconds(1));
+
+            cpuFilterTimer.async_wait([&](const boost::system::error_code& ec) {
+                if (ec == boost::asio::error::operation_aborted)
+                {
+                    /* we were canceled*/
+                    return;
+                }
+                if (ec)
+                {
+                    std::cerr << "timer error\n";
+                    return;
+                }
+                createSensors(io, objectServer, systemBus, nullptr, false);
+            });
+        };
+
     std::vector<std::unique_ptr<sdbusplus::bus::match_t>> matches =
         setupPropertiesChangedMatches(*systemBus, sensorTypes, eventHandler);

@@ -1265,6 +1320,11 @@ int main()
             "',arg0namespace='xyz.openbmc_project.Inventory.Item'",
         cpuPresenceHandler));

+    matches.emplace_back(std::make_unique<sdbusplus::bus::match_t>(
+        static_cast<sdbusplus::bus_t&>(*systemBus),
+        sdbusplus::bus::match::rules::interfacesAdded(inventoryPath),
+        addCpuPresenceHandler));
+
     getPresentCpus(systemBus);

     setupManufacturingModeMatch(*systemBus);
--
2.43.0

